File: leet_LIST_1721_swap_nodes_from_k.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


# solved again in 2023

class Solution:
    def swapNodes(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        p = head

        while p and k > 1:
            p = p.next
            # i += 1
            k -= 1
        from_left = p
        from_right = head
        while p.next:
            from_right = from_right.next
            p = p.next
        from_left.val, from_right.val = from_right.val, from_left.val
        return head

"""
class Solution:
    def swapNodes(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        lend = head
        rend = head
        p = head
        while k > 1:
            lend = lend.next
            p = p.next
            k -= 1
        while p.next:
            rend = rend.next
            p = p.next
        n = lend.val
        lend.val = rend.val
        rend.val = n
        return head
"""

File: leet_LIST_1721_swap_nodes_from_k.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* swapNodes(ListNode* head, int k)
    {
        ListNode    *p, *L, *R;

        L = R = p = head;
        
        while (k != 1)
        {
            p = p->next;
            L = L->next;
            --k;
        }
        while (p->next)
        {
            p = p->next;
            R = R->next;
        }

        int tmp = L->val;
        L->val = R->val;
        R->val = tmp;
        return head;
    }
};
